packages/webpack/lib/utils/resolvable.js

```diff
@@ -17,28 +19,33 @@ exports.Resolvable = class Resolvable {
         });
       }
     };
-    Object.assign(this, {
-      reset() {
-        state.splice(0, state.length);
-      },
-      resolve(value) {
-        Promise.resolve(value).then(
-          (value) => queue.forEach((fn) => fn(null, value)),
-          this.reject
-        );
-      },
-      reject(reason) {
-        queue.forEach((fn) => fn(reason));
-      },
-      then(onFulfilled, onRejected) {
-        return promise().then(onFulfilled, onRejected);
-      },
-      catch(onRejected) {
-        return promise().catch(onRejected);
-      },
-    });
-    if (executor) {
-      new Promise(executor).then(this.resolve, this.reject);
-    }
+    this.reset = (executor) => {
+      state.splice(0, state.length);
+      if (executor) {
+        new Promise((resolve, reject) =>
+          executor(
+            (value) => void resolve(value) || this.resolve(value),
+            (error) => void reject(error) || this.reject(error),
+            (executor) => this.reset(executor)
+          )
+        ).catch(this.reject);
+      }
+    };
+    this.resolve = (value) => {
+      Promise.resolve(value).then(
+        (value) => queue.slice().forEach((fn) => fn(null, value)),
+        this.reject
```

Comment:
Why do we need this additional promise? And this `reject` function will never be called as the promise is already resolved.

Reply:
I am afraid we do: `value` could be a `Promise`. That is the reason I added `this.reset` here, too.

Generally, I want this thing to mimick the behavior of a real `Promise`.